### PR TITLE
backend/feat:forward batch

### PR DESCRIPTION
**Pull request by aayushagarwal2001**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable
- [ ] No leak detected in leakcanary

### CODE REVIEW

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPool.hs [hunk 1]
```diff
@@ -119,20 +131,22 @@ prepareDriverPoolBatch driverPoolCfg searchReq searchTry tripQuoteDetails starti
...
-      (currentDriverPoolBatch, poolType, nextScheduleTime) <-
-        calculateWithFallback poolTypesWithFallback $ \poolType -> withLogTag (show poolType <> "_step=" <> show radiusStep) $ do
+      prepareDriverPoolBatchEntity <-
+        calculateWithFallback poolTypesWithFallback $ \poolType -> do
           allDriversNotOnRide <- calcDriverPool poolType radiusStep merchantOpCityId_
           case poolType of
             SkipPool -> do
               incrementBatchNum searchTry.id
-              prepareDriverPoolBatch' previousBatchesDrivers (batchNum + 1) True merchantOpCityId_ txnId
+              (batch, poolType', nextScheduleTime) <- prepareDriverPoolBatch' previousBatchesDrivers (batchNum + 1) True merchantOpCityId_ txnId
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPool.hs:148]: is function ko hi sundar krde bhai

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPool.hs [hunk 2]
```diff
@@ -162,34 +176,38 @@ prepareDriverPoolBatch driverPoolCfg searchReq searchTry tripQuoteDetails starti
...
+              pure $ PrepareDriverPoolBatchEntity goHomePoolPostCalc poolType (Just goHomeConfig.goHomeBatchDelay) []
             NormalPool -> do
               let allNearbyDriversCurrentlyNotOnRide = filterSpecialDrivers transporterConfig.specialDrivers allDriversNotOnRide
-              allNearbyDriversCurrentlyOnRide <- calcDriverCurrentlyOnRidePool poolType radiusStep transporterConfig merchantOpCityId_
-              (,poolType,Nothing) <$> calculateNormalBatch merchantOpCityId_ transporterConfig intelligentPoolConfig (allNearbyDriversCurrentlyOnRide <> allNearbyDriversCurrentlyNotOnRide) radiusStep blockListedDrivers txnId
-      logDebug $ "finalPool: " <> show currentDriverPoolBatch
-      logDebug $ "finalPoolType: " <> show poolType
-      cacheBatch currentDriverPoolBatch
-      let poolWithMicroBatching =
+              (normalBatchNotOnRide, normalBatchOnRide) <- calculateNormalBatch merchantOpCityId_ transporterConfig intelligentPoolConfig allNearbyDriversCurrentlyNotOnRide radiusStep blockListedDrivers txnId poolType batchNum False
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPool.hs:194]: batchNum pehle se hai

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide/Internal.hs [hunk 3]
```diff
@@ -146,7 +146,9 @@ cancelRideTransaction ::
 cancelRideTransaction booking ride bookingCReason merchantId rideEndedBy = do
   let driverId = cast ride.driverId
   void $ CQDGR.setDriverGoHomeIsOnRideStatus ride.driverId booking.merchantOperatingCityId False
-  QDI.updateOnRide False (cast ride.driverId)
+  activeAdvancedRide <- QRide.getActiveAdvancedRideByDriverId ride.driverId
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/CancelRide/Internal.hs:149]: This seems to be a common function. Lets have it at one place.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPool/Config.hs [hunk 4]
```diff
@@ -34,7 +36,23 @@ data BatchSplitByPickupDistance = BatchSplitByPickupDistance
   deriving stock (Show, Eq, Read, Ord, Generic)
   deriving anyclass (FromJSON, ToJSON, ToSchema)
 
+data BatchSplitByPickupDistanceOnRide = BatchSplitByPickupDistanceOnRide
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPool/Config.hs:39]: Lets keep it as `BatchSplitByPickupDistance`

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPool.hs
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPool.hs:1]: Revisit this file again.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Ride.hs [hunk 5]
```diff
@@ -78,15 +79,16 @@ initializeRide merchantId driver booking mbOtpCode enableFrequentLocationUpdates
             pure otpCode
           Just otp -> pure otp
   ghrId <- CQDGR.setDriverGoHomeIsOnRideStatus driver.id booking.merchantOperatingCityId True
-
-  ride <- buildRide driver booking ghrId otpCode enableFrequentLocationUpdates mbClientId
+  previousRideInprogress <- QRide.getInProgressByDriverId driver.id
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Ride.hs:82]: Lets revisit stuck case with @sidsethu

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/RideExtra.hs [hunk 6]
```diff
@@ -182,6 +182,17 @@ findOneByDriverId (Id personId) = findAllWithOptionsKV [Se.Is BeamR.driverId $ S
 getInProgressByDriverId :: (MonadFlow m, EsqDBFlow m r, CacheFlow m r) => Id Person -> m (Maybe Ride)
 getInProgressByDriverId (Id personId) = findOneWithKV [Se.And [Se.Is BeamR.driverId $ Se.Eq personId, Se.Is BeamR.status $ Se.Eq Ride.INPROGRESS]]
 
+getInProgressByDriverIds :: (MonadFlow m, EsqDBFlow m r, CacheFlow m r) => [Id Person] -> m [(Ride, DTC.TripCategory)]
+getInProgressByDriverIds driverIds = do
+  rides <- findAllWithKV [Se.And [Se.Is BeamR.driverId $ Se.In $ getId <$> driverIds, Se.Is BeamR.status $ Se.Eq Ride.INPROGRESS]]
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/RideExtra.hs:187]: Think about it.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/RideExtra.hs [hunk 7]
```diff
@@ -194,15 +205,14 @@ getActiveByDriverId (Id personId) =
...
     ( \ride ->
         maybeM
-          (return [])
-          (\booking -> return [(ride, booking)])
+          (return Nothing)
+          (\booking -> return $ Just (ride, booking))
           (QBooking.findById ride.bookingId)
     )
-    (findOneWithKV [Se.And [Se.Is BeamR.driverId $ Se.Eq personId, Se.Is BeamR.status $ Se.In [Ride.INPROGRESS, Ride.NEW]]])
+    =<< findAllWithKV [Se.And [Se.Is BeamR.driverId $ Se.Eq personId, Se.Is BeamR.status $ Se.In [Ride.INPROGRESS, Ride.NEW]]]
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/RideExtra.hs:215]: Lets discuss if we can avoid DB calls here

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/SearchRequestForDriver.hs [hunk 8]
```diff
@@ -132,12 +132,14 @@ instance FromTType' BeamSRFD.SearchRequestForDriver SearchRequestForDriver where
             createdAt = T.localTimeToUTC T.utc createdAt,
             vehicleServiceTier = fromMaybe (castVariantToServiceTier vehicleVariant) vehicleServiceTier,
             goHomeRequestId = Id <$> goHomeRequestId,
+            rideFrequencyScore = rideFrequencyScore,
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/SearchRequestForDriver.hs:135]: ??

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/spec/Storage/configs.yaml [hunk 9]
```diff
@@ -49,13 +58,19 @@ DriverPoolConfig:
...
 
   default:
     tripCategory: "'All'"
     area: "'Default'"
     distanceBasedBatchSplit: "Array ['BatchSplitByPickupDistance { batchSplitSize = 1, batchSplitDelay = 0 }', 'BatchSplitByPickupDistance { batchSplitSize = 1, batchSplitDelay = 4 }']"
     scheduleTryTimes: "'{1800, 900, 300}'"
+    batchSizeOnRide: "10"
+    onRideRadiusConfig: "ARRAY[]::TEXT[]"
+    currentRideTripCategoryValidForForwardBatching: "Array ['OneWay_OneWayOnDemandDynamicOffer']"
+    onRideBatchSplitConfig: "Array ['BatchSplitByPickupDistance { batchSplitSize = 1, batchSplitDelay = 0 }', 'BatchSplitByPickupDistance { batchSplitSize = 1, batchSplitDelay = 4 }']"
```
[the remark this answers is not in the document]
Comment (hkmangla) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/spec/Storage/configs.yaml:73]: Lets fix it in Next PR!

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs [hunk 10]
```diff
@@ -907,8 +908,8 @@ respondQuote (driverId, merchantId, merchantOpCityId) clientId mbBundleVersion m
     searchReq <- QSR.findById searchTry.requestId >>= fromMaybeM (SearchRequestNotFound searchTry.requestId.getId)
     merchant <- CQM.findById searchReq.providerId >>= fromMaybeM (MerchantDoesNotExist searchReq.providerId.getId)
     driver <- QPerson.findById driverId >>= fromMaybeM (PersonNotFound driverId.getId)
-    driverInfo <- QDriverInformation.findById (cast driverId) >>= fromMaybeM DriverInfoNotFound
-    when driverInfo.onRide $ throwError DriverOnRide
+    driverInfo <- QDriverInformation.findById driverId >>= fromMaybeM (PersonNotFound driverId.getId)
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs:911]: Dude! Fix in next PR.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride.hs [hunk 11]
```diff
@@ -317,6 +324,7 @@ mkDriverRideRes rideDetails driverNumber rideRating mbExophone (ride, booking) b
         nextStopLocation = nextStopLocation,
         lastStopLocation = lastStopLocation,
         tripScheduledAt = booking.startTime,
+        bookingType = if ride.status == DRide.NEW && ride.isAdvanceBooking && maybe False (.hasAdvanceBooking) driverInfo then ADVANCED else CURRENT,
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride.hs:327]: mention the reason in comment.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/EndRide/Internal.hs [hunk 12]
```diff
@@ -129,7 +129,8 @@ endRideTransaction ::
   TransporterConfig ->
   m ()
 endRideTransaction driverId booking ride mbFareParams mbRiderDetailsId newFareParams thresholdConfig = do
-  QDI.updateOnRide False (cast ride.driverId)
+  updateOnRideStatusWithAdvancedRideCheck ride.driverId
+  QDI.updateHasAdvancedRide (cast ride.driverId) False
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Ride/EndRide/Internal.hs:133]: Check it again!

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Transformers/Ride.hs [hunk 13]
```diff
@@ -70,3 +71,8 @@ mkOdometerReading fileId odoValue = odoValue <&> (\value -> OdometerReading (Id
 
 mkLatLong :: Maybe Double -> Maybe Double -> Maybe LatLong
 mkLatLong lat lon = LatLong <$> lat <*> lon
+
+getTripCategory :: Maybe Common.TripCategory -> Common.TripCategory
+getTripCategory tripCategory = case tripCategory of
+  Just cat -> cat
+  Nothing -> Common.OneWay Common.OneWayOnDemandDynamicOffer
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Transformers/Ride.hs:78]: Handle it properly. Take from Booking.

**Review by hkmangla — APPROVED** [review]

**Pull request merged**